src/dynamicprompts/samplers/combinatorial.py

```diff
@@ -32,136 +30,113 @@ def _dedupe(arr: list[str]) -> tuple[str, ...]:
         self,
-        *,
-        wildcard_manager: WildcardManager,
-        ignore_whitespace: bool = False,
-        sampler_router: SamplerRouter,
-        parser_config: ParserConfig = default_parser_config,
-    ):
-        super().__init__(
-            wildcard_manager=wildcard_manager,
-            ignore_whitespace=ignore_whitespace,
-            sampler_router=sampler_router,
-            parser_config=parser_config,
-        )
-        self._sampler_router = sampler_router
-
-    def _propagate_sampling_method(self, commands: Iterable[Command]) -> None:
-        for cmd in commands:
-            if cmd.sampling_method == SamplingMethod.DEFAULT:
-                cmd.sampling_method = SamplingMethod.COMBINATORIAL
-
-    def _get_sequence(self, command: SequenceCommand) -> StringGen:
-        self._propagate_sampling_method(command.tokens)
-
-        sentinel_start = LiteralCommand(
-            "",
-            sampling_method=SamplingMethod.COMBINATORIAL,
-        )
-        sentinel_end = LiteralCommand("", sampling_method=SamplingMethod.COMBINATORIAL)
-
+        command: SequenceCommand,
+        context: SamplingContext,
+    ) -> StringGen:
         non_combo_commands = [
             c
             for c in command.tokens
             if c.sampling_method != SamplingMethod.COMBINATORIAL
         ]
         command_collection = CommandCollection(
             non_combo_commands,
-            self._sampler_router,
+            context=context,
         )
-        command.tokens.insert(0, sentinel_start)
-        command.tokens.append(sentinel_end)
+        augmented_tokens = [
+            LiteralCommand(
+                "",
+                sampling_method=SamplingMethod.COMBINATORIAL,
+            ),  # sentinel 1
+            *command.tokens,
+            LiteralCommand(
+                "",
+                sampling_method=SamplingMethod.COMBINATORIAL,
+            ),  # sentinel 2
+        ]
 
         def get_sequence(commands: list[Command]) -> typing.Iterable[list[str]]:
             if len(commands) == 0:
                 yield []
             else:
                 first_command, rest = commands[0], commands[1:]
-                if first_command.sampling_method != SamplingMethod.COMBINATORIAL:
+                if context.get_effective_sampling_method(first_command).is_nonfinite():
                     for rest_vals in get_sequence(rest):
                         val = command_collection.get_value(first_command)
                         if val:
                             yield [val] + rest_vals
                         else:
                             yield rest_vals
                 else:
-                    gen = self._sampler_router.generator_from_command(first_command)
+                    gen = context.generator_from_command(first_command)
                     for first_val in gen:
                         for rest_vals in get_sequence(rest):
                             yield [first_val] + rest_vals
```

Comment:
As with `_combo_to_prompt` above, the same argument applies. The different ways of handling finite and nonfinite commands should be made more explicit to help with readability. For consistency, we may want to use the CommandManager in `_combo_to_prompt` as well.

perhaps something like

```python
def sample_with_nonfinite_command(nonfinite_command: Command, other_commands: list[Commands]) -> Iterable[list_of_tokens]:
    for token_list in get_sequence(other_commands):
        val = command_collection.get_value(nonfinite_command)  or "" # or .sample_command(nonfinite_command)
        yield [val] + token_list

def sample_with_finite_command(finite_command: Command, other_commands: list[Commands]) -> Iterable[list_of_tokens]:
    token_sampler = context.generator_from_command(finite_command)
    for token in token_sampler:
        for token_list in get_sequence(other_commands):
            yield [token] + token_list

sample_with_nonfinite_command = context.get_effective_sampling_method(first_command).is_nonfinite()
if sample_with_nonfinite_command:
    yield from sample_with_nonfinite_command(first_command, rest)
else:
    yield from sample_with_finite_command(first_command, rest)
```